homeassistant/components/modem_callerid/sensor.py

```diff
@@ -28,33 +25,48 @@
     }
 )
 
+SERVICE_SCHEMA = vol.Schema({vol.Optional(ATTR_ENTITY_ID): cv.comp_entity_ids})
+
+_LOGGER = logging.getLogger(__name__)
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up modem caller ID sensor platform."""
 
-    name = config.get(CONF_NAME)
-    port = config.get(CONF_DEVICE)
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the Modem Caller ID sensor."""
+    name = entry.data[CONF_NAME]
+    device = entry.data[CONF_DEVICE]
+    api = hass.data[DOMAIN][entry.entry_id][DATA_KEY_API]
+    async_add_entities(
+        [
+            ModemCalleridSensor(
+                hass,
+                api,
+                name,
+                device,
+                entry.entry_id,
+            )
+        ]
+    )
 
-    modem = bm(port)
-    if modem.state == modem.STATE_FAILED:
-        _LOGGER.error("Unable to initialize modem")
-        return
+    platform = entity_platform.current_platform.get()
 
-    add_entities([ModemCalleridSensor(hass, name, port, modem)])
+    platform.async_register_entity_service(
+        SERVICE_REJECT_CALL, SERVICE_SCHEMA, "reject_call"
+    )
 
 
 class ModemCalleridSensor(SensorEntity):
     """Implementation of USB modem caller ID sensor."""
 
-    def __init__(self, hass, name, port, modem):
+    def __init__(self, hass, api, name, device, server_unique_id):
         """Initialize the sensor."""
         self._attributes = {"cid_time": 0, "cid_number": "", "cid_name": ""}
-        self._name = name
-        self.port = port
-        self.modem = modem
+        self._device_class = None
+        self.device = device
+        self.api = api
         self._state = STATE_IDLE
-        modem.registercallback(self._incomingcallcallback)
-        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self._stop_modem)
+        self._name = name
+        self._server_unique_id = server_unique_id
+        api.registercallback(self._incomingcallcallback)
```

Comment:
Please move side effects out of `__init__`

Reply:
Do you mean ```api.registercallback(self._incomingcallcallback)```? If so, where should I put it or should it be ```self.api.registercallback(self._incomingcallcallback)```?

Reply:
Move it to a setup function and call it after creating the object

Reply:
Move it to a setup function and call it after creating the object

Reply:
So far, I've been unsuccessful with this. The kira integration also seems to use this in ```__init__```.

Reply:
Potential solution https://github.com/home-assistant/core/pull/46677/files#r619883169